### ISSUE TITLE
Using pnpm

### ISSUE THREAD

**Issue by aminya** [atom-ide-community/atom#61]
pnpm is much faster than `npm` or `yarn`. We should consider using that to speed up the build process. It is totally compatible with `npm` so switching to it is just a matter of changing `npm` to `pnpm` in the scripts that install something.

https://pnpm.js.org/en/benchmark

![image](https://cdn.rawgit.com/pnpm/node-package-manager-benchmark/7f0c3e40/results/imgs/react-app.svg)

**Comment by DeeDeeG** [comment]
Mixed feelings:

First of all, neat idea.

Second, I am concerned this would be moved forward too quickly without being vetted. If we test it enough and know it inside and out, and can clearly prove that various points of different behavior from vanilla `npm` are going to be okay, then sure.

- I think we should avoid using `pnpm`'s `pnpm-lock.yaml` if `pnpm` will let us. (If it will let us just keep using `package-lock.json` as a lockfile, but still install faster in CI.) Remember, we maintain all diff from upstream ourselves. One less custom piece of different tech, one less bit of maintenance/bugfix/chore work to worry about. We don't want to spend time with this fork doing that kind of work, IMO.
- > **Strict.** A package can access only dependencies that are specified in its package.json.
  - Will this break any packages? Spend the time to make sure it doesn't. Packages might be relying `npm`'s looser behavior here.

Regarding `apm`: neat idea, but I don't want to actually do this.

Let's not do this in `apm` bundled with Atom unless it has been supremely well-tested. Core infrastructure like `apm` is responsible for building Atom's bundled packages, and any user/community packages. And speaking as someone who's patched it, it's a pile of spaghetti. Surprisingly easy to break.

`apm` is an elaborate wrapper around vanilla `npm`. Using it with `pnpm` instead could take a large rewrite of the codebase in order to still work as designed. Any quirks in implementation of either `npm` or `pnpm` could derail this. Would not be surprised if `apm` requires features of `npm` that aren't in `pnpm`. Not looking forward to verifying that _none_ of `apm` breaks with `pnpm`, would be quite surprised if so.

Given that `apm` is a pile of spaghetti that mostly works, trying to rewrite it is going to be unpleasant. Better to make it simpler and _more_ standard than try to improve it in a non-standard way.

**Comment by DeeDeeG** [comment]
I will be happy with this if:
- It installs faster

But I will be unhappy with this if:
- It breaks things in Atom
- It makes us heavily Patch Atom, or fork and patch other areas of the Atom ecosystem, to work with `pnpm`

**Comment by DeeDeeG** [comment]
Steps to switch to pnpm:

- install `pnpm` in `script/package[-lock].json`.
- Make a `getPnpmBinPath()` function in `script/config.js` based on `getNpmBinPath()`.
- Switch to `pnpm` in `script/lib/install-script-dependencies.js` with the new `getPnpmBinPath()` function from `config.js`
  - Make sure this works!
- Switch to `pnpm` in `script/lib/install-apm.js` just like above
  - Make sure this works!
- Well, that's it until/unless we get it working in `apm`.
  - Leave this until the above is verified twenty times over! Vet up front or you will spend many times as long debugging later.

**Comment by DeeDeeG** [comment]
Issue to resolve: `pnpm`doesn't support the `--global-style` flag Atom uses to install `apm`.

(The flag changes how `node_modules` are installed, reducing deduping, and potentially changing which modules/version of modules are reachable from a given package).

Is this flag necessary in upstream? We should independently validate that dropping that flag is okay against upstream's `master`, as not otherwise modified for this fork (minimal tweaks to CI to allow CI to run are acceptable).

**Comment by aminya** [comment]
Certainly, we should test this. But if we can make it work for atom, that is a good improvement! 

I have started to use pnpm for other packages of this organization. It works without any issues. Check these for example:
https://github.com/atom-ide-community/atom-ide-javascript
https://github.com/atom-ide-community/atom-ide-datatip
https://github.com/atom-ide-community/eslint-config-atomic

Regarding the lock file, I do not think it is a good idea to keep the original package-lock. If we decide to switch, we should do it completely. I always keep the `lock update` commits separate from the rest, so I can drop them in case of a rebase. If we do so, there will not be any issues.

**Comment by ThatXliner** [comment]
> * I think we should avoid using `pnpm`'s `pnpm-lock.yaml` if `pnpm` will let us. (If it will let us just keep using `package-lock.json` as a lockfile, but still install faster in CI.) Remember, we maintain all diff from upstream ourselves. One less custom piece of different tech, one less bit of maintenance/bugfix/chore work to worry about. We don't want to spend time with this fork doing that kind of work, IMO.

Not a problem anymore 😄
